File: python_files/raylib/pong.py
import pyray as p

SCREEN_WIDTH: int = 800
SCREEN_HEIGHT: int = 600
SCREEN_TITLE: str = "PONG"
SCREEN_BACKGROUND: p.Color = p.LIGHTGRAY


class Ball:
    def __init__(self) -> None:
        self.initial_x: float = p.get_screen_width()/2
        self.initial_y: float = p.get_screen_height()/2
        self.x: float = self.initial_x
        self.y: float = self.initial_y
        self.radius: float = 10.0
        self.initial_speed: float = 300.0
        self.speed: float = self.initial_speed
        self.speed_increment: float = 10.0
        self.direction = p.Vector2(
            -1 if p.get_random_value(0, 1) == 0 else 1,
            -1 if p.get_random_value(0, 1) == 0 else 1,
        )
        self.color: p.Color = p.RED
        self.is_active: bool = False

    def reset(self) -> None:
        self.x = self.initial_x
        self.y = self.initial_y
        self.speed = self.initial_speed
        self.direction = p.Vector2(
            -1 if p.get_random_value(0, 1) == 0 else 1,
            -1 if p.get_random_value(0, 1) == 0 else 1,
        )
        self.is_active = False

    def draw(self) -> None:
        p.draw_circle_v(p.Vector2(self.x, self.y), self.radius, self.color)

    def update(self) -> None:
        # activate/deactivate ball
        if p.is_key_pressed(p.KeyboardKey.KEY_SPACE):
            self.is_active = not self.is_active

        # move ball
        if self.is_active:
            self.x += self.direction.x * self.speed * p.get_frame_time()
            self.y += self.direction.y * self.speed * p.get_frame_time()

        # normalize direction vector
        if self.direction.x != 0 and self.direction.y != 0:
            self.direction = p.vector2_normalize(self.direction)

        # ball bounds
        if self.y < self.radius or self.y > p.get_screen_height() - self.radius:
            self.direction.y *= -1


class Paddle:
    def __init__(self, x: float) -> None:
        self.width: float = 10.0
        self.height: float = 100.0
        self.initial_x: float = x
        self.initial_y: float = p.get_screen_height()/2 - self.height/2
        self.x: float = self.initial_x
        self.y: float = self.initial_y
        self.speed: float = 300.0
        self.color: p.Color = p.BLACK

    def reset(self) -> None:
        self.x = self.initial_x
        self.y = self.initial_y

    def draw(self) -> None:
        p.draw_rectangle_rec(p.Rectangle(
            self.x, self.y, self.width, self.height), self.color)

    def update_player(self) -> None:
        if p.is_key_down(p.KeyboardKey.KEY_UP) and self.y > 0:
            self.y -= self.speed * p.get_frame_time()
        if p.is_key_down(p.KeyboardKey.KEY_DOWN) and self.y < p.get_screen_height() - self.height:
            self.y += self.speed * p.get_frame_time()

    def update_ai(self, ball_y: float) -> None:
        if ball_y < self.y + self.height/2 and self.y > 0:
            self.y -= self.speed * p.get_frame_time()
        if ball_y > self.y + self.height/2 and self.y < p.get_screen_height() - self.height:
            self.y += self.speed * p.get_frame_time()


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    score_left: int = 0
    score_right: int = 0

    ball: Ball = Ball()
    player: Paddle = Paddle(10.0)
    ai: Paddle = Paddle(p.get_screen_width() - player.width - 10.0)

    while not p.window_should_close():

        ball.update()
        player.update_player()
        ai.update_ai(ball.y)

        # ball collision  paddle
        if p.check_collision_circle_rec(p.Vector2(ball.x, ball.y), ball.radius, p.Rectangle(player.x, player.y, player.width, player.height)):
            ball.direction.x *= -1
            ball.speed += ball.speed_increment
        if p.check_collision_circle_rec(p.Vector2(ball.x, ball.y), ball.radius, p.Rectangle(ai.x, ai.y, ai.width, ai.height)):
            ball.direction.x *= -1
            ball.speed += ball.speed_increment

        # update scores
        if ball.x < ball.radius:
            score_right += 1
            ball.reset()
            player.reset()
            ai.reset()
        if ball.x > p.get_screen_width() - ball.radius:
            score_left += 1
            ball.reset()
            player.reset()
            ai.reset()

        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        # draw scores
        p.draw_text(str(score_left), 200, 30, 40, p.DARKGRAY)
        p.draw_text(str(score_right), p.get_screen_width() -
                    200, 30, 40, p.DARKGRAY)

        ball.draw()
        player.draw()
        ai.draw()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: cpp_files/raylib/pong.cpp
#include <raylib.h>
#include <raymath.h>

class Ball
{
  public:
    Vector2 initial_pos;
    Vector2 pos;
    Vector2 direction;
    float radius;
    float initial_speed;
    float speed;
    float speed_increment;
    Color color;
    bool is_active;

    Ball()
    {
        initial_pos = (Vector2){ GetScreenWidth() * 0.5f, GetScreenHeight() * 0.5f };
        pos = initial_pos;
        direction = (Vector2){
            static_cast<float>(GetRandomValue(0, 1) == 0 ? -1 : 1),
            static_cast<float>(GetRandomValue(0, 1) == 0 ? -1 : 1),
        };
        radius = 10.0f;
        initial_speed = 300.0f;
        speed = initial_speed;
        speed_increment = 10.0f;
        color = RED;
        is_active = false;
    }

    void reset()
    {
        pos = initial_pos;
        direction = (Vector2){
            static_cast<float>(GetRandomValue(0, 1) == 0 ? -1 : 1),
            static_cast<float>(GetRandomValue(0, 1) == 0 ? -1 : 1),
        };
        speed = initial_speed;
        is_active = false;
    }

    void draw() const
    {
        DrawCircleV(pos, radius, color);
    }

    void update()
    {
        // activate/deactivate ball
        if (IsKeyPressed(KEY_SPACE))
        {
            is_active = !is_active;
        }

        // move ball
        if (is_active)
        {
            pos.x += direction.x * speed * GetFrameTime();
            pos.y += direction.y * speed * GetFrameTime();
        }

        // normalize direction vector
        if (direction.x != 0 && direction.y != 0)
        {
            direction = Vector2Normalize(direction);
        }

        // ball bounds
        if (pos.y < radius || pos.y > GetScreenHeight() - radius)
        {
            direction.y *= -1;
        }
    }
};

class Paddle
{
  public:
    Rectangle initial_rect;
    Rectangle rect;
    float speed;
    Color color;

    Paddle(float x)
    {
        float height = 100.0f;
        initial_rect =
          (Rectangle){ x, static_cast<float>(GetScreenHeight()) / 2 - height / 2, 10.0f, height };
        rect = initial_rect;
        speed = 300.0f;
        color = BLACK;
    }

    void reset()
    {
        rect = initial_rect;
    }

    void draw() const
    {
        DrawRectangleRec(rect, color);
    }

    void update_player()
    {
        if (IsKeyDown(KEY_UP) && rect.y > 0)
        {
            rect.y -= speed * GetFrameTime();
        }
        if (IsKeyDown(KEY_DOWN) && rect.y < GetScreenHeight() - rect.height)
        {
            rect.y += speed * GetFrameTime();
        }
    }

    void update_ai(float ball_y)
    {
        if (ball_y < rect.y + rect.height / 2 && rect.y > 0)
        {
            rect.y -= speed * GetFrameTime();
        }
        if (ball_y > rect.y + rect.height / 2 && rect.y < GetScreenHeight() - rect.height)
        {
            rect.y += speed * GetFrameTime();
        }
    }
};

int
main(void)
{
    const int SCREEN_WIDTH = 800;
    const int SCREEN_HEIGHT = 600;
    const char SCREEN_TITLE[] = "PONG";
    const Color SCREEN_BACKGROUND = LIGHTGRAY;

    InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

    int score_left = 0;
    int score_right = 0;

    Ball ball;
    Paddle player(10.0f);
    Paddle ai(GetScreenWidth() - player.rect.width - 10.0f);

    while (!WindowShouldClose())
    {
        ball.update();
        player.update_player();
        ai.update_ai(ball.pos.y);

        // ball collision paddle
        if (CheckCollisionCircleRec(ball.pos, ball.radius, player.rect))
        {
            ball.direction.x *= -1;
            ball.speed += ball.speed_increment;
        }
        if (CheckCollisionCircleRec(ball.pos, ball.radius, ai.rect))
        {
            ball.direction.x *= -1;
            ball.speed += ball.speed_increment;
        }

        // update scores
        if (ball.pos.x < ball.radius)
        {
            score_right++;
            ball.reset();
            player.reset();
            ai.reset();
        }
        if (ball.pos.x > GetScreenWidth() - ball.radius)
        {
            score_left++;
            ball.reset();
            player.reset();
            ai.reset();
        }

        BeginDrawing();
        ClearBackground(SCREEN_BACKGROUND);

        // draw scores
        DrawText(TextFormat("%d", score_left), 200, 30, 40, DARKGRAY);
        DrawText(TextFormat("%d", score_right), GetScreenWidth() - 200, 30, 40, DARKGRAY);

        ball.draw();
        player.draw();
        ai.draw();

        EndDrawing();
    }

    CloseWindow();

    return 0;
}